### PR TITLE
feat(provider/kubernetes): Add a new k8s client java api client.

### PR DESCRIPTION
**Pull request by chlung**
We prefer small, well tested pull requests.

Please refer to [Contributing to Spinnaker](https://spinnaker.io/community/contributing/).

When filling out a pull request, please consider the following:

* Follow the commit message conventions [found here](http://www.spinnaker.io/v1.0/docs/how-to-submit-a-patch).
* Provide a descriptive summary for your changes.
* If it fixes a bug or resolves a feature request, be sure to link to that issue.
* Add inline code comments to changes that might not be obvious.
* Squash your commits as you keep adding changes.
* Add a comment to @spinnaker/reviewers for review if your issue has been outstanding for more than 3 days.

Note that we are unlikely to accept pull requests that add features without prior discussion. The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

This PR will add a new k8s java client to leverage these newly support api from kubernetes client api library.  Outline my changes
- Create a client and use the passing in config file to setup connection configuration.  Some additional or missing configuration will be added based on feedback.
- Create a new method loadStatefulSets() in KubernetesSecurityGroupCachingAgent  to ensure that the data can be retrieve using this new client and api.
- We do include these k8s client jar files in this commit this time.  Once the jar published to maven repo and we shall move them.

This PR will prepare us to next PR for CachingAgent changes for supporting these new controllers.

Thanks.

### CODE REVIEW

**Comment by chlung** [comment]
@lwander Could you review and provide your feedback.  Especially, the client configuration part.  You might know something use case sthat I haven't covered.  Thanks.

File: clouddriver-kubernetes/clouddriver-kubernetes.gradle [hunk 1]
```diff
@@ -5,4 +5,7 @@ dependencies {
   compile spinnaker.dependency('frigga')
   compile spinnaker.dependency('bootActuator')
   compile spinnaker.dependency('bootWeb')
+
+  //FIXME: Remove the following line when k8s client java api jar is available in maven repo
+  compile fileTree(dir: 'kubernetes-client-java-libes', include: '*.jar')
```
Comment (lwander) [inline clouddriver-kubernetes/clouddriver-kubernetes.gradle]: Is "libes" a typo of "libs"?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy [hunk 2]
```diff
@@ -214,6 +225,17 @@ class KubernetesApiAdaptor {
     }
   }
 
+  List<V1beta1StatefulSet> getStatefulSets(String namespace) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy]: This is not relevant to this PR

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy [hunk 3]
```diff
@@ -0,0 +1,112 @@
+package com.netflix.spinnaker.clouddriver.kubernetes.security
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: Missing license header

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+  public ApiClient parseConfig(String fileName) throws IOException {
+    return this.parseConfig((Reader)(new FileReader(fileName)));
+  }
+
+  public ApiClient parseConfig(InputStream stream) {
+    return parseConfig((Reader)(new InputStreamReader(stream)));
+  }
+
+  public ApiClient parseConfig(Reader input) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: This feels like functionality handled by the util package, no? Or is this done to handle the extra config properties in the kubernetes-clouddriver configuration?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/provider/agent/KubernetesServerGroupCachingAgent.groovy [hunk 5]
```diff
@@ -197,6 +198,13 @@ class KubernetesServerGroupCachingAgent extends KubernetesCachingAgent implement
     }.flatten()
   }
 
+  List<ReplicaSet> loadStatefulSets() {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/provider/agent/KubernetesServerGroupCachingAgent.groovy]: Why is this returning replica sets?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesNamedAccountCredentials.groovy [hunk 6]
```diff
@@ -213,7 +213,9 @@ public class KubernetesNamedAccountCredentials implements AccountCredentials<Kub
       }
 
       return new KubernetesCredentials(
-          new KubernetesApiAdaptor(name, config, spectatorRegistry),
+          new KubernetesApiAdaptor(name, config,
+            new KubernetesApiClientConfig(kubeconfigFile, context, cluster, user, namespaces, serviceAccount),
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesNamedAccountCredentials.groovy]: I'd pull this constructor invocation onto a separate line, this is becoming a little dense/hard to parse

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy [hunk 7]
```diff
@@ -41,18 +46,24 @@ class KubernetesApiAdaptor {
...
+    if (!config || !clientConfig) {
       throw new IllegalArgumentException("Config may not be null.")
     }
     this.config = config
     this.account = account
     this.client = new DefaultKubernetesClient(this.config)
     this.spectatorRegistry = spectatorRegistry
     this.spectatorClock = spectatorRegistry.clock()
+
+    apiClient = clientConfig.getApiCient()
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy]: I think we're going to have an easier time by not conflating both api clients in the same adapter, but to instead use 2 separate adapters.

File: clouddriver-kubernetes/clouddriver-kubernetes.gradle (hunk 1 above)
Comment (chlung) [inline clouddriver-kubernetes/clouddriver-kubernetes.gradle]: This is a typo.  I will fix it.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy (hunk 7 above)
Comment (chlung) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy]: You wan tot have a new file for the client?

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy (hunk 2 above)
Comment (chlung) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesApiAdaptor.groovy]: This is to prove the client is fully function as expected.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy (hunk 3 above)
Comment (chlung) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: ok. I will fix this problem.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy (hunk 4 above)
Comment (chlung) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: I try to us the util.Config, but I don't know if we are going to handle any additional fields or not.  Also, I am not sure if we need apiVersion, namespace, ... which we discuss it before.  I can change just use util.Config.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesNamedAccountCredentials.groovy (hunk 6 above)
Comment (chlung) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesNamedAccountCredentials.groovy]: ok.  I can fix this one.

**Comment by chlung** [comment]
@lwander I just want to confirm about the following items.
1.  Do you want to have a new adapter file for the new client?
2. Should we just use util.Config package instead of create a new one?  

Thanks

**Comment by lwander** [comment]
1. Yes that's cleanest
2. Yes

**Comment by chlung** [comment]
@lwander I made those changes per your feedback.  Please review again.  About loadStatefulSets() code, we  will move into the new caching agent when we implement the new code.  Thanks.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy [hunk 8]
```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2016 Google, Inc.
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy]: This should probably be 2017 Cisco, not google

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    account ? new KubernetesOperationException(account, "$operation in $namespace", e) :
+      new KubernetesOperationException("$operation in $namespace", e)
+  }
+
+  KubernetesOperationException formatException(String operation, ApiException e) {
+    account ? new KubernetesOperationException(account, "$operation", e) :
+      new KubernetesOperationException("$operation", e)
+  }
+
+  Boolean blockUntilResourceConsistent(Object desired, Closure<Long> getGeneration, Closure getResource) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy]: Are these copied directly from the other Adaptor? If so, they should probably be pulled into a separate class and be shared.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+      wait = [wait * 2, RETRY_MAX_WAIT_MILLIS].min()
+
+      current = getResource()
+    }
+
+    return true
+  }
+
+  /*
+   * Atomically create a new client, and pass it to the given doOperation closure to operate against the kubernetes API
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy]: This comment is probably relevant to the other client's creation, not this one

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy [hunk 11]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2016 Google, Inc.
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: License header needs to be updated again

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy (hunk 9 above)
Comment (chlung) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy]: The code are not exactly same.  One is using fabric8 exception and another is use 8s client api exception.

Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/api/KubernetesClientApiAdapter.groovy]: OK that's fair

**Comment by lwander** [comment]
Looks good - ready to merge?

**Comment by chlung** [comment]
yes.  thanks.
